core/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -151,9 +147,17 @@ public ExecutorScalingQueue() {
 
         @Override
         public boolean offer(E e) {
+            // first try to transfer to a waiting worker thread
             if (!tryTransfer(e)) {
+                // check if there might be spare capacity in the thread
+                // pool executor
```

Comment:
I don't think this is what is happening here. AFAICS the expression `executor.getMaximumPoolSize() - executor.getCorePoolSize()` is fixed once we have constructed the threadpool and only changed by `setCorePoolSize()` and `setMaximumPoolSize()`.

What I think this code effectively does is to behave like a fixed size thread pool with unbounded queue in case max pool size is configured to be equal to core pool size (which is what @bleskes was essentially saying, I guess). If that's the case, I also concur with @bleskes that we should get rid of this case distinction. This means that we should require min < max for scaling threadpool and otherwise require to explicitly specify fixed thread pool.